sphinx/io.py

```diff
@@ -153,30 +152,45 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
 
 
-def read_doc(app: "Sphinx", env: BuildEnvironment, filename: str) -> nodes.document:
+def read_doc(publisher: Publisher, docname: str, filename: str) -> nodes.document:
     """Parse a document and convert to doctree."""
     # set up error_handler for the target document
-    error_handler = UnicodeDecodeErrorHandler(env.docname)
+    error_handler = UnicodeDecodeErrorHandler(docname)
     codecs.register_error('sphinx', error_handler)  # type: ignore
 
+    publisher.set_source(source_path=filename)
+    publisher.publish()
+
+    doctree = publisher.document
+    # settings get modified in ``write_doctree``; get a local copy
```

Comment:
This mentions to the behavior of `env.write_doctree()`. It's better to move this to `env.read_doc()`.

BTW, what is the merit of cloning?